sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1042,6 +1042,13 @@ void ProgramManager::addImages(pi_device_binaries DeviceBinary) {
         auto Result = KSIdMap.insert(std::make_pair(EntriesIt->name, KSId));
         (void)Result;
         assert(Result.second && "Kernel sets are not disjoint");
+        // ... and create a unique kernel ID for the entry
+        std::shared_ptr<detail::kernel_id_impl> KernelIDImpl =
+            std::make_shared<detail::kernel_id_impl>(EntriesIt->name);
+        sycl::kernel_id KernelID =
+            detail::createSyclObjFromImpl<sycl::kernel_id>(KernelIDImpl);
+        m_KernelIDs.insert(
+            std::make_pair(EntriesIt->name, std::move(KernelID)));
```

Comment:
Can we have a diagnostic if a kernel name is already defined(aka multiple definitions)?
IIRC this function is usually called from a function with `__attirbute__((constructor))`, so throwing an exception might not make sense. Maybe there is another good place where we can have this diagnostic?

Reply:
> Maybe there is another good place where we can have this diagnostic?

Probably, right before the very first submit is the right place. Though, this verification should be properly guarded an be thread-safe.

Reply:
I believe the disjunction check [above](https://github.com/intel/llvm/pull/4379/files/6ca4defe7a408be793c0ff45b9d8ac6d791aa345#diff-78dd7f7ba0b6120dece1ae4ab5a09c9936ff654a1de2c31ff2dbb1fc58d90393R1044) should guarantee that there are no duplicates. That is unless there are duplicates between OS modules. I don't know if duplicates between OS modules are valid, but if they are, would it be sufficient to let them share kernel IDs?

Reply:
Duplicate kernel names from different OS modules are not valid.
End user won't have the check "above" since it's an assert. Probably we can add a diagnostic(rt check + throw) somewhere on line 1326.